Correct client area origin and navigation back from secondary wxFrame on IOS

RobertRoeb opened the pull request:
- Add extra space for iPhone status bar, currently hard coded as a temporary fix until we can calculate it
  - Use it in GetClientAreaOrigin() and use that, respectively, in wxTopLevelWindow::Layout to place the client window correctly
  - Add default wxToolbar in all secondary frames as they have no close button with an [X], but most usually a < like sign as a way to return to the previous wxFrame = "screen"

Commit:
Correct client area pos and navigation back from secondary wxFrame on iOS

  - Add extra space for iPhone status bar, currently hard coded as
    a temporary fix until we can calculate it
  - Use it in GetClientAreaOrigin() and use that, respectively,
    in wxTopLevelWindow::Layout to place the client window correctly
  - Add default wxToolbar in all secondary frames as they have no
    close button with an [X], but most usually a < like sign as a
    way to return to the previous wxFrame = "screen"

vadz commented on art/prev_screen.xpm:
```diff
@@ -0,0 +1,121 @@
+/* XPM */
```
We should be really using PNGs for any new icons IMNSHO.

RobertRoeb replied:
Not sure why- all icons that we currently ship are XPM

vadz commented on src/common/toplvcmn.cpp:
```diff
@@ -496,7 +496,13 @@ bool wxTopLevelWindowBase::Layout()
             DoGetClientSize(&clientW, &clientH);
 
-            child->SetSize(0, 0, clientW, clientH);
+#ifndef __WXOSX_IPHONE__
```
Can we please avoid this, either by using `GetClientAreaOrigin()` under all platforms if this is indeed the right thing to do or by adding another virtual function, e.g. `GetUsableAreaOrigin()` if it is not. But using such preprocessor checks is just not a good idea.

vadz commented on src/osx/carbon/frame.cpp:
```diff
@@ -60,6 +66,20 @@ bool wxFrame::Create(wxWindow *parent,
         return false;
 
+#ifdef __WXOSX_IPHONE__
+    if (parent != NULL) {
```
```suggestion
    if (parent != nullptr)
    {
```

RobertRoeb replied:
Done

vadz commented on src/osx/carbon/frame.cpp:
```diff
@@ -60,6 +66,20 @@ bool wxFrame::Create(wxWindow *parent,
+        tb->AddTool( wxID_CLOSE, wxEmptyString, wxBitmap( prev_screen_xpm ) );
+        tb->AddStretchableSpace();  // doesn't yet stretch on iOS, so at least add two of them
+        tb->AddStretchableSpace();
+        tb->AddControl( new wxStaticText( tb, -1, title ) );
```
```suggestion
        tb->AddControl( new wxStaticText( tb, wxID_ANY, title ) );
```

vadz commented on src/osx/carbon/frame.cpp:
```diff
@@ -60,6 +66,20 @@ bool wxFrame::Create(wxWindow *parent,
+    if (parent != NULL) {
+        // We are on the next screen, provide a back button and title
+        wxToolBar *tb = CreateToolBar();
+        // tb->AddTool( wxID_CLOSE, wxEmptyString, wxArtProvider::GetBitmap(wxART_PREV_SCREEN) ); doesn't work for some reason
```
It really should, but if it doesn't, it would be worth opening an issue about it and removing this commented out line.

vadz commented on src/osx/carbon/frame.cpp:
```diff
@@ -60,6 +66,20 @@ bool wxFrame::Create(wxWindow *parent,
+        wxToolBar *tb = CreateToolBar();
+        // tb->AddTool( wxID_CLOSE, wxEmptyString, wxArtProvider::GetBitmap(wxART_PREV_SCREEN) ); doesn't work for some reason
+        tb->AddTool( wxID_CLOSE, wxEmptyString, wxBitmap( prev_screen_xpm ) );
+        tb->AddStretchableSpace();  // doesn't yet stretch on iOS, so at least add two of them
```
Again, instead of working around this here, why not make `AddStretchableSpace()` in wxiOS implementation add 2 spaces?

RobertRoeb replied:
Done

vadz commented on src/osx/carbon/frame.cpp:
```diff
@@ -397,8 +417,22 @@ void wxFrame::PositionToolBar()
         else
         {
 #if !wxOSX_USE_NATIVE_TOOLBAR
+#ifdef __WXOSX_IPHONE__
```
This should also be a virtual function in `wxWidgetImpl`.

vadz commented on src/osx/nonownedwnd_osx.cpp:
```diff
@@ -253,6 +253,20 @@ wxPoint wxNonOwnedWindow::GetClientAreaOrigin() const
     int left, top, width, height;
     m_nowpeer->GetContentArea(left, top, width, height);
+
+#ifdef __WXOSX_IPHONE__
```
And by having it in a virtual function we'd at least avoid duplicating this.

vadz commented on src/osx/nonownedwnd_osx.cpp:
```diff
@@ -470,6 +484,17 @@ void wxNonOwnedWindow::DoGetClientSize( int *width, int *height ) const
 #ifdef __WXOSX_IPHONE__
     GetPeer()->GetContentArea(left, top, w, h);
+
+    // TODO, find a way to find the safe area not covered at the top
```
Sorry, triplicating.

vadz reviewed (changes requested):
Sorry but this is really more of a draft-quality than a ready-to-merge PR. Could you please polish it at least a little bit? TIA!

Commit:
Streamlined GetClientAreaOrigin(), enable stretchable toolbar items in IOS

  - Implemented stretchable toolbar items in iOS. This was difficult as
    code in the common toobar code assumes the tools are created later
    in wxToolbar::Realize() which is not the case for iOS. I added a
    wxITEM_STRETCHABLE_SEPARATOR instead of writing new overloads for
    all CreateTool()s in all ports with a bool stretchable=false parameter
  - Streamlined the GetClientAreaOrigin() code moved the code that
    queries the size of the iOS status bar into src/iphone
  - This might have broken something

Commit:
Use nullptr instead of NULL

RobertRoeb commented:
I tested the changes on wxMac and wxGTK and didn't see anything broken. Indeed, I reverted all or nearly all changes in non-OSX or non-IOS code.
<img width="424" height="905" alt="BackNavi IOS" src="https://github.com/user-attachments/assets/7905d52c-2126-4c5b-86ca-ea200b3bc7e9" />

RobertRoeb commented:
Here is my hack over a hack to allow stretchable spacer in iOS toolbar. The other solution would be to remove both hacks and add a flag to CreateTool() in all ports. I didn't do that because someone must have thought about that when he added the hack that is currently in the code and decided against the clean solution. 

BTW, I tried really hard to also use Realize() by removing the unstretchable spacers in the toolbar and inserting stretchable ones if they were marked with MarkStretchable(). That _almost_ worked but ruined the layout of the toolbar for reasons I could not figure out after 2h.

virtual wxToolBarToolBase *CreateTool(int toolid,
                                          const wxString& label,
                                          const wxBitmapBundle& bmpNormal,
                                          const wxBitmapBundle& bmpDisabled = wxBitmapBundle(),
                                          wxItemKind kind = wxITEM_NORMAL,
                                          wxObject *clientData = nullptr,
                                          const wxString& shortHelp = wxEmptyString,
                                          const wxString& longHelp = wxEmptyString,
                                          **bool stretchable = false** ) = 0;


<<<< This is in the PR right now

wxToolBarToolBase *wxToolBarBase::InsertStretchableSpace(size_t pos)
{
#ifdef __WXOSX_IPHONE__
   // The hack below does not work on iPhone as the tools are
   // created in the constructor not Realize(). I hack around the hack
   // by adding a distinct ID
   const int id = wxID_STRETCHABLE_SEPARATOR;
#else
   const int id = wxID_SEPARATOR;
#endif
    wxToolBarToolBase * const tool = CreateSeparator( id );
    if ( tool )
    {
        // this is a hack but we know that all the current implementations
        // don't really use the tool when it's created, they will do it
        // InsertTool() at earliest and maybe even in Realize() much later
        //
        // so we can create the tool as a plain separator and mark it as being
        // a stretchable space later
        tool->MakeStretchable();
    }

    return DoInsertNewTool(pos, tool);
}

Commit:
Corrected typos and whitespace issues

Commit:
Add check for wxCAPTION/wxCLOSE_BOX before adding to iOS toolbar

csomor commented:
> The "clean" solution is not really clean because `bool stretchable` only makes sense when combined with `wxITEM_SEPARATOR` and nothing else. A decent API shouldn't allow specifying nonsensical combinations of parameters, like `wxITEM_CHECK` and `stretchable=true`. Of course, `wxToolBar` is already guilty of it, but it's not a reason to make things worse.
> 
> The only alternative API I see is to add `wxITEM_STRETCHABLE_SEPARATOR` but this doesn't seem appealing. Making an existing separator stretchable doesn't seem very out of place as it's similar to making an existing check tool checked etc. Of course, if this _really_ can't be implemented for iOS we may want to revisit it, but for now I guess we can live with `wxID_STRETCHABLE_SEPARATOR` hack, although I think making `CreateSeparator()` virtual and overriding it in wxiOS `wxToolBar` would be slightly less ugly than checking for iOS in common code.

I'll try to move things into a separate PR. @vadz what about making `MakeStretchable` virtual ? I think I could recreate the native item.

vadz commented:
Sure, making `MakeStretchable()` virtual would be perfectly fine. TIA!

csomor commented:
As the different parts have been merged individually I'll close this soon if nobody objects
